mmscan/evaluator/metrics/box_metric.py

```diff
@@ -0,0 +1,277 @@
+    """
+    confidences = subset_results['scores']
+    sample_indices = subset_results['sample_indices']
+    ious = subset_results['ious']
+    gt_matched_records = {}
+    total_gt_boxes = 0
+    for i, sample_idx in enumerate(sample_indices):
+        if sample_idx not in gt_matched_records:
+            gt_matched_records[sample_idx] = np.zeros((len(ious[i]), ),
+                                                      dtype=bool)
+            total_gt_boxes += ious[i].shape[0]
+
+    confidences = np.array(confidences)
+    sorted_inds = np.argsort(-confidences)
+    sample_indices = [sample_indices[i] for i in sorted_inds]
+    ious = [ious[i] for i in sorted_inds]
+
+    tp_thr = np.zeros(len(sample_indices))
+    fp_thr = np.zeros(len(sample_indices))
+
+    for d, sample_idx in enumerate(sample_indices):
+        iou_max = -np.inf
+        cur_iou = ious[d]
+        num_gts = cur_iou.shape[0]
+        if num_gts > 0:
+            for j in range(num_gts):
+                iou = cur_iou[j]
+                if iou > iou_max:
+                    iou_max = iou
+                    jmax = j
+
+        if iou_max >= iou_thr:
+            if not gt_matched_records[sample_idx][jmax]:
+                gt_matched_records[sample_idx][jmax] = True
+                tp_thr[d] = 1.0
+            else:
+                fp_thr[d] = 1.0
+        else:
+            fp_thr[d] = 1.0
+
+    fp = np.cumsum(fp_thr)
+    tp = np.cumsum(tp_thr)
+    recall = tp / float(total_gt_boxes)
+    precision = tp / np.maximum(tp + fp, np.finfo(np.float64).eps)
+
+    return average_precision(recall, precision), np.max(recall)
+
+
+def get_average_precision(iou_array: np.ndarray, iou_threshold: float) \
+        -> Tuple[np.ndarray, np.ndarray]:
+    """Return the average precision and max recall for a given iou array.
+
+    Args:
+        iou_array (ndarray/tensor):
+            The iou matrix of the predictions and ground truths
+            (shape len(preds)*len(gts))
+        iou_threshold (float): 0.25/0.5
+
+    Returns:
+        Tuple[np.ndarray, np.ndarray]: the average precision and max recall.
+    """
+    fp, tp = __get_fp_tp_array__(iou_array, iou_threshold)
+    fp_cum = np.cumsum(fp)
+    tp_cum = np.cumsum(tp)
+    recall = tp_cum / float(iou_array.shape[1])
+    precision = tp_cum / np.maximum(tp_cum + fp_cum, np.finfo(np.float64).eps)
+
+    return average_precision(recall, precision), np.max(recall)
+
+
+def get_multi_topk_scores(iou_array: Union[np.ndarray, torch.tensor],
+                          iou_threshold: float,
+                          mode: str = 'sigma') -> Dict[str, float]:
+    """
+    Compute the multi-topk metric, we provide two modes.
+    "simple": 1/N *Hit(min(N*k,len(pred)))
+    "sigma": 1/N * Sigma [Hit(min(n*k,len(pred)))>=n] n = 1~N
+        Hit(M) return the number of gtound truths hitted by the first
+        M predictions.
+        N = the number of gtound truths
```

Comment:
can move the explanation about different modes to the docstring about the "argument mode"